### ISSUE TITLE
Editor UI freezes when editing a particular class

### ISSUE THREAD

**Issue by oleg-okhrimenko** [mplushnikov/lombok-intellij-plugin#516]
### Short description
Editing the class causes UI to freeze for some noticeable time. While Editor is blocked high CPU load is observed. The thread dump of IDEA is attached.
[threadDumps-freeze-20180731-123943-IU-181.5281.24-101sec.zip](https://github.com/mplushnikov/lombok-intellij-plugin/files/2262007/threadDumps-freeze-20180731-123943-IU-181.5281.24-101sec.zip)
The feeze happens when calling code completion, pasting something or performing other actions causing some syntax highlight change, compile error displaying, etc.
Found a reference for similar issue on IDEA's bug tracker:
https://youtrack.jetbrains.com/issue/KT-25031
Proposed workaround (disabling plugin) indeed allows to unlock the editor. Although, as far as we intensively use Lombok features, that's not an option.
Class causing the problem
[BriefDTO.java.zip](https://github.com/mplushnikov/lombok-intellij-plugin/files/2262020/BriefDTO.java.zip)
The issue is reproduced on every team member machine (some running under Windows, some - under MacOS)

### Expected behavior
Expected to work smoothly.

### Version information

 * **IDEA Version:** IntelliJ IDEA 2018.1.5 (Ultimate Edition); Build #IU-181.5281.24, built on June 12, 2018
 * **JDK Version:** _JRE: 1.8.0_152-release-1136-b39 x86_64
 * **OS Type & Version:** I AM RUNNING MacOS 10.13.6, although the issue gets reproduced on machines running Windows 8.1 and Windows 10 as well.
 * **Lombok Plugin Version:** _0.19-2018.1
 * **Lombok Dependency Version:** 1.18.0

### Steps to reproduce
 1. Open class for editing
 2. Try to call code completion or paste something

### Stacktrace
Stack traces are not in IDEA log, but inside of special thread dumps generated. I have attached one of such generated dumps compressed.

**Comment by vrosenshtein** [comment]
I have the same problem - reproduced on my Windows desktop running on core-i5 and 16GB of RAM.

**Comment by greenposer** [comment]
Same issue on my Windows desktop with same IDE and Lombok versions.

**Comment by mplushnikov** [comment]
@nicity Thank you for looking at this issue. I'm using only FileBasedIndex access in this method and have hoped it's fast enougth already... Do you have some points for good caching examples inside of IntelliJ to look for?

**Comment by ewirch** [comment]
@nicity The Youtrack issue does not exist any more.

**Comment by ewirch** [comment]
If you need more info to reproduce the problem, see attached sample project.

Copy these lines from `Test2` class:
```
         mapper = new CsvMapper();
         mapper.setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE);
         schema = CsvSchema.emptySchema().withHeader()
                 .withColumnSeparator(';');
```

And paste them into `my.pckge.Test.test()`. There is a significant delay of 2s. The delay goes away if you remove the Lombok class annotations. Or if you remove the fields.

[test-project.zip](https://github.com/mplushnikov/lombok-intellij-plugin/files/2436903/test-project.zip)

**Comment by nicity** [comment]
@mplushnikov Please take a look for CachedValue / CachedValuesManager
@ewirch The issue has private visibility as of now

**Comment by mplushnikov** [comment]
@oleg-okhrimenko / @ewirch  Do you use "lombok.config" in your projects?
I cann't see any delays in your example @ewirch ... What is Idea/plugin versions, you are using?

@nicity I'm already using CachedValuesManager in AugmentProvider with dependency to PsiModificationTracker.JAVA_STRUCTURE_MODIFICATION_COUNT. What can I used as dependency for FileBasedIndex operations?

**Comment by oleg-okhrimenko** [comment]
@mplushnikov, yes we use it, although it holds only a single line: "config.stopBubbling = true". Just when revisiting the issue, I have came to one conclusion. Before the issue we were not using Lombok's @FieldNameConstants annotation. If i comment this out on a class, then restart IDEA and issue code completion - it takes less time to show completion option then when it is on.

**Comment by ewirch** [comment]
The significant delay went away after upgrade to IntelliJ 183.3283.2 and Lombok plugin 0.21.IDEA-EAP. There is still some delay, but it does not depend on the Lombok annotations any more. I can trigger it with pojo classes. So I guess it is not a Lombok plugin problem. At least not for me.

**Comment by oleg-okhrimenko** [comment]
[lombok-freeze-demo.zip](https://github.com/mplushnikov/lombok-intellij-plugin/files/2445387/lombok-freeze-demo.zip)
It seems that I was able to reproduce the issue in simple project. Attached archive with it. In order to reproduce you need to:
1. Unpack archive
2. Open as idea project
3. Issue gradle refresh
4. Navigate to main/java source set and open the only class called Brief DTO
5. Opening you should experience IDEA being hung as well as freeze dumps start to appear in idea logs folder.

Please note, that if we remove unnecessary here static import issue is either gone or feels better.

Hope this helps

**Comment by steven509** [comment]
I have a class which would completely freeze Inteliji if I made any changes in it. Since updating to the same versions as @ewirch it only freezes a little bit which is a lot more manageable.

**Comment by mplushnikov** [comment]
@oleg-okhrimenko Thank you very much, I can reproduce some performance degradation with your example project!

**Comment by FatihErdem** [comment]
I have a class which has 25 fields and annotated with `@Builder` annotation. It freezes when I tried to add other annotations to fields. I faced with a huge performance loss.


#### Setup
IntelliJ IDEA 2018.2.4 (Ultimate Edition)
Build #IU-182.4505.22, built on September 18, 2018
Lombok version: 1.18.2
Lombok Plugin version: Version: 2.0.7
JRE: 1.8.0_152-release-1248-b8 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.13.6

**Comment by mplushnikov** [comment]
Just added something, that could help to get better performance. Can you guys check this version in your environment?
[lombok-plugin-0.22-2018.2.zip](https://github.com/mplushnikov/lombok-intellij-plugin/files/2451893/lombok-plugin-0.22-2018.2.zip)

**Comment by FatihErdem** [comment]
@mplushnikov I will try on Monday.

**Comment by ewirch** [comment]
You limited the IntelliJ compatibility by purpose to `until-build="182.*"`?. Can't load it, since I'm on 183.* EAP.

**Comment by mplushnikov** [comment]
@oleg-okhrimenko Thank you very much for testing!

@ewirch Yes, sorry, I have separate build for EAP version:
[lombok-plugin-0.22-183.2153.8.zip](https://github.com/mplushnikov/lombok-intellij-plugin/files/2456744/lombok-plugin-0.22-183.2153.8.zip)

**Comment by ewirch** [comment]
Let's put it this way: its usable using Lombok plugin 0.21 and 0.22. 0.22 did not add any significant change. Adding a single line break in a class with many fields, triggers activity which calms down only after 5s. Here's a screen cast of activity monitor during me adding a line break:

![load](https://user-images.githubusercontent.com/1880345/46669122-cc888280-cbce-11e8-8383-9f85abe8227d.gif)

If I remove all Lombok annotations from the class, the system takes similarly long to calm down after a new line break. Strangely Lombok plugin is still involved in CPU consumption.

I can't tell if this normal behavior or not. At least the UI is not blocked. Those are background activities. I can continue to work.

**Comment by mplushnikov** [comment]
@htfv thank you for testng. I will release this version soon.

@ewirch Is it still reproducable in your example project (https://github.com/mplushnikov/lombok-intellij-plugin/issues/516#issuecomment-426178762)? Maybe you can find some other example for me?

**Comment by ewirch** [comment]
No, not reproducible any more. There is still a slight delay when pasting lines. But that's nothing compared to the original problem.

**Comment by htfv** [comment]
> No, not reproducible any more. There is still a slight delay when pasting lines. But that's nothing compared to the original problem.

That't true, it does not feel very smooth, but there are no real freezes. We had several-second freezes on every save before this fix.

**Comment by mplushnikov** [comment]
Ok, I close it for now. Feel free to open new ticket.
